[Draft] Add function body implementation in opschema

raymondxyang opened the pull request:
* Add function_body field in opschema, add verification for function op register
* Make type/shape inference optional for opschema with function body
* Add related python APIs and cpp test case

Commit:
Add function body implementation in onnx opschema

raymondxyang commented:
@gramalingam could you also help take a look? Thanks

Commit:
Merge with master

Commit:
Fix

Commit:
Fix

Commit:
Fix copy constructor and gcc warning

Commit:
Fix

Commit:
Update doc genration scripts to contain function information

Commit:
Sync with master branch

Commit:
Update

Commit:
Flake

Commit:
typecheck

Commit:
Fix header

Commit:
Add support type

Commit:
Rename testcase

Commit:
Update documents

gramalingam commented on onnx/checker.cc:
```diff
@@ -304,6 +304,131 @@ void check_attribute(
+    int domain_version) {
+  if (!schema || schema->Deprecated()) {
+    // There's no primitive operator for the node.
+    // Check whether it's referring to a function.
```
Is this an out-dated comment? I don't understand it.

raymondxyang replied:
It is indeed a outdated comment. Removed

gramalingam commented on onnx/checker.cc:
```diff
@@ -304,6 +304,131 @@ void check_attribute(
+    // There's no primitive operator for the node.
+    // Check whether it's referring to a function.
+    fail_check(
+        "No Op registered for " + node.op_type() + " with domain_version of " +
```
Wouldn't a different error message be appropriate for a "deprecated op"?

Commit:
Fix draft comments

Commit:
Sync with master

Commit:
Update doc

gramalingam commented on onnx/defs/schema.cc:
```diff
@@ -20,6 +20,48 @@ void RegisterSchema(OpSchema&& schema) {
+  ctx.set_opset_imports(op_set);
+  ctx.set_is_main_graph(false);
+  
+  checker::TENSOR_TYPES_MAP tc_map;
```
This type checking part is more complex. I suggest removing this from the current PR into a separate one. There are two challenges. One is that if the function inputs are allowed to have one of several different types, then the checking becomes more complex. The simplest way to do that is to check for every possible type assignment to all variables. Secondly, input.GetTypes() may return a type-variable name instead of a type.

gramalingam commented on onnx/checker.cc:
```diff
@@ -304,6 +304,133 @@ void check_attribute(
+  }
+  verify_node_schema(node, schema, domain_version);
+
+  // Type constraints check between op and function body.
```
I suggest removing this check from the current PR. There are a couple of issues here. First, the "type strings" being compared here may be type-variables, and not types. Second, this check is incomplete, since it does not do type inference for all nodes, and checks only direct references to function parameters.

gramalingam commented on onnx/defs/schema.cc:
```diff
@@ -775,6 +823,19 @@ OpName_Domain_Version_Schema_Map& OpSchemaRegistry::map() {
+      for (auto& opset_entry_ : m) {
+        for (auto& op_versions_entry_ : opset_entry_.second) {
+          for (auto& op_entry_ : op_versions_entry_.second) {
+            VerifyFunction(op_entry_.second, m);
```
Hi, why not do the call to VerifyFunction as an extra step in "Finalize" (around line 690 in old file and line 740 in your version)? I think that will be better.

Commit:
Move function verify test to CI

Commit:
Merge branch 'master' of github.com:onnx/onnx into function-op

gramalingam commented:
Hi Raymond, the changes look good. I have a question/suggestion: there seems to some code duplication between check_function and check_graph. Is it worth eliminating this duplication, may be using a common shared function called from both?

gramalingam commented:
Another question: is VerifyFunctionNode used anywhere? Now, it seems to have no usage inside ONNX. (Expanding the function definition and checking it is required only for type-checking; so, it seems unnecessary for the checking done here.)

gramalingam commented on onnx/test/cpp/function_verify_test.cc:
```diff
@@ -0,0 +1,151 @@
+      }
+      for (auto& type : tensor_name_tc.second) {
+        if (allowed_types.find(type) == allowed_types.end()) {
+          fail_check(
```
We can't call this an error yet, because the elements in allowed_types might be a type variable. Furthermore, the two separate loops (lines 53 to 56 and the one in line 57 is not required, since we should be able to directly do the check at line 55 for the ith argument.

raymondxyang replied:
Hi Rama.. sorry just saw your comment. see L22-L29. I exclude the type variable already. This one is only enforced on primitive types

raymondxyang replied:
For the other two comments: the check_function is a existing function. I didnt change any of that (just misplaced it so there is diff.) Ill put it back and see if we can do any modification on that. For VerifyFunctionNode Ill remove it. Thanks for pointing that out 😃

raymondxyang replied:
L23 should be all_tensor_types(). Ill change it.

gramalingam replied:
Hi Raymond, you excluded it in tc_map. But allowed_types can include type-variables as well. You have not excluded it from that. In any case, I don't think that the overall structure for this type-checking is the right one for extending it to handle all cases.

raymondxyang replied:
Sorry I am not familiar with the type variable: If a node in a function node, which consumes or produce a tensor with variable type, and is also a exposed tensor of the whole function op, should this variable type also appear in tc_map, which generated from the overall defined type constraints?

raymondxyang replied:
I try to figure out a way to check constaints between op's type variable and node inside function body but somehow cannot find a specific rule in order to generate one.. I think we can explore how to resolve this later.. its a huge item

gramalingam replied:
Sorry, ignore my comment about type-variables. I see that "Finalize" on OpSchema would have converted the type-variable to a list of proper types using type-constraints already. (But I still think we will need to replace this with the logic we have in "Graph::InferAndVerifyTypeMatch(…)" etc.)

gramalingam replied:
But we can do that separately. Can you just add a comment line saying "TODO: apply complete type inference across function body for a more complete check?"

raymondxyang replied:
Sure. I already put a todo there.

Commit:
Fix and removing unused code

Commit:
Fix

Commit:
Remove redundant field

postrational committed:
More extendable Runner (#1809)

* Make Runner class more flexible

* Change @staticmethods to @classmethods in test Runner

* Rename assert_similar_outputs and download_model functions

Commit:
Add missed changes in proto files

Commit:
Merge branch 'master' into function-op

Commit:
fix

Commit:
Update __init__.py

Commit:
Revert "Remove redundant field"

This reverts commit cdbe6b420d4f4230860d33c83c93e815f9fa56d3.

Commit:
Revert "Add missed changes in proto files"

This reverts commit 88826887b3b927b9a4619da38a58f51c2b2c6b9b.

liwchang committed:
Fix shape inference when auto_pad is notset again (#1830)

Treat auto_pad notset as no auto_pad,
since our newest Modelzoo onnx models still make auto_pad notset instead of removing it.

Commit:
Resolve comments

Commit:
Merge branch 'function-op' of github.com:raymondxyang/onnx into function-op

Commit:
Merge branch 'master' into function-op

Commit:
Fix gcc

gramalingam commented on onnx/defs/experiments/experiments_functions.cc:
```diff
@@ -9,161 +9,155 @@ namespace ONNX_NAMESPACE {
-  func.set_status(OperatorStatus::STABLE);
-
-  NodeProto* initial_node0 = func.add_node();
+static std::vector<NodeProto> BuildMVN() {
```
I am slightly concerned that there may be some extra copying going on here: but may be the vectors will be "moved" cheaply by the compiler (avoiding redundant copies)?

gramalingam commented on onnx/defs/schema.cc:
```diff
@@ -622,6 +622,38 @@ void OpSchema::ParseAndSetTypes(
   }
 }
 
+OpSchema& OpSchema::FunctionBody(const std::vector<NodeProto>& func_nodes) {
```
Can we avoid copies if we use a non-reference parameter "std::vector<NodeProto> func_nodes" and do a std::move? Just wondering what is the best way to avoid some of the needless copies of nodes.

Commit:
Resolve comments on doc

Commit:
Resolve comments

Commit:
Resolve comments

Commit:
Fix const reference

Commit:
Renaming

linkerzhang committed:
Remove exp op: Affine, ImageScaler,ParametricSoftplus, Crop. (#1832)

* Remove exp op: Affine, ImageScaler,ParametricSoftplus, Crop.

* update the test case.

hariharans29 committed:
Change TopK operator to allow dynamic 'k' (#1829)

* More changes

* More changes

* More changes

* More changes

* Check-in modified model for topK

* More changes

* More changes

* More cahnges

* Changes to md files

* More changes

* More changes

* Fix broken build

* Fix build break

* Fix build break

* Fix build break

* More changes

* Fix build break

* revert

* revert file to old state

Commit:
Resolve PR comments

Commit:
Rename function

linkerzhang committed:
bump the version for drop out - fix the issue that the version was not bumped when changing its type constraint declaration. (#1848)

houseroad committed:
remove global variable in header file (#1850)

Commit:
Resolve PR comments

yuslepukhin committed:
Handle string comparision represented as np.objects (#1851)

and all other objects.

linkerzhang committed:
Merge branch 'master' into function-op
